_posts/2020-12-01-the-road-to-the-jit.md

```diff
@@ -0,0 +1,382 @@
+was less optimized than LLVM-generated code because it would still use
+BEAM's stack and X registers (stored in memory), but the overhead for
+[instruction unpacking and instruction dispatch][overhead] was
+eliminated.
+
+[overhead]: http://blog.erlang.org/a-closer-look-at-the-interpreter/
+
+The initial benchmarks results were promising: about twice as fast
+compared to interpreted BEAM code, so Lukas extended the prototype
+so that it could handle more kind of BEAM instructions.
+
+[John Högberg][john] quickly became interested in the project and
+started to act as a sounding board.  Some time later, probably in
+March, John suggested that the new JIT should translate **all** loaded
+code to native code.  That way, there would be no need to support
+context switching between the BEAM interpreter and native code, which
+would make the design simpler and eliminate the cost for context
+switches.
+
+That was a gamble, of course. After all, it could turn out that the
+native code could be too large to be practically useful or decrease
+performance because it fitted badly in the code cache.  They decided
+that it was worth taking the risk and that it would probably be
+possible to optimize the size of the code later.  (*Spoiler*: At the time
+of writing, the native code generated by the JIT is about 10 percent
+larger than interpreted BEAM code.)
+
+Another change to the design was the tooling for generating the native
+code. In Lukas's prototype, the native code template for each
+instruction was contained in text files similar to the other files
+used by the loader.  That was inflexible, so it was decided to use
+some library that could generate native code.  While some pure C
+libraries could have been used, the C++ library AsmJIT was more
+convenient in practical use than any of the C libraries.  Therefore
+the part of the loader that translates BEAM instructions to native
+code needed to be written in C++, but the rest of the runtime system
+is still pure C code and will remain so.
+
+John joined the practical work on the rejigged JIT project at the end
+of March.
+
+On April 7, 2020, John reached the "prompt beer" milestone.
+
+### Prompt Beer
+
+When the Erlang system is started, a surprisingly large amount of code is
+executed before the prompt appears. On the one hand that means that
+the translation of many instructions needs to be implemented
+before it would be possible to even start the Erlang system, let alone run
+any test suites or benchmarks.
+
+On the other hand, when the prompt finally appears, it is a major
+milestone worth celebrating with some prompt beer or other appropriate
+beverage or by taking the rest of the evening off.
+
+### Maturing the new JIT
+
+On April 14 John got Dialyzer running with the JIT, and on April 17
+after some improvements to the code generation Dialyzer was only about
+10 percent slower with the JIT than with HiPE. (At the time of
```

Comment:
"after some improvements to the code generation" => ", after some improvements to the code generation,"